aten/src/ATen/native/cuda/MaxUnpooling.cu

```diff
@@ -0,0 +1,653 @@
+
+namespace at {
+namespace native {
+
+template <typename T>
+__host__ __device__ __forceinline__ T ceilDiv(T a, T b) {
+  return (a + b - 1) / b;
+}
+
+template <typename T>
+__global__ void MaxUnpooling2d_forward_kernel(
+    const int64_t numInputElements,
+    const T* input,
+    const int64_t* indices,
+    const int64_t numBatch,
+    const int64_t numChannels,
+    const int64_t inputHeight,
+    const int64_t inputWidth,
+    const int64_t outputHeight,
+    const int64_t outputWidth,
+    T* output) {
+  for (int linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < numInputElements;
+       linearIndex += blockDim.x * gridDim.x) {
+    int c = (linearIndex / inputWidth / inputHeight) % numChannels;
+    int n = linearIndex / inputWidth / inputHeight / numChannels;
+    output += (n * numChannels + c) * outputHeight * outputWidth;
+    int maxind = indices[linearIndex];
+    output[maxind] = input[linearIndex];
+  }
+}
+
+template <typename T>
+__global__ void MaxUnpooling3d_forward_kernel(
+    const T* input,
+    const int64_t* indices,
+    const int64_t batchSize,
+    const int64_t inputSlices,
+    const int64_t inputTime,
+    const int64_t inputHeight,
+    const int64_t inputWidth,
+    const int64_t oT,
+    const int64_t oW,
+    const int64_t oH,
+    const int64_t dT,
+    const int64_t dW,
+    const int64_t dH,
+    const int64_t pT,
+    const int64_t pW,
+    const int64_t pH,
+    const int64_t offsetZ,
+    T* output) {
+  int64_t iColumn = blockIdx.x * blockDim.x + threadIdx.x;
+  int64_t iRow = blockIdx.y * blockDim.y + threadIdx.y;
+  int64_t iFrame = (blockIdx.z + offsetZ) % inputTime; // intput frame/time
+  int64_t slice = (blockIdx.z + offsetZ) / inputTime; // intput slice/feature
+  if (iRow < inputHeight && iColumn < inputWidth) {
+    int64_t newIndex = slice * (inputTime * inputHeight * inputWidth) +
+        iFrame * (inputHeight * inputWidth) + iRow * inputWidth + iColumn;
+    T val = input[newIndex];
+    int64_t index = indices[newIndex];
+    output[slice * oT * oH * oW + index] = val;
+  }
+}
+
+template <typename T>
+__global__ void MaxUnpooling2d_backward_kernel(
+    const int64_t numInputElements,
+    const T* input,
+    const int64_t* indices,
+    const int64_t numBatch,
+    const int64_t numChannels,
+    const int64_t inputHeight,
+    const int64_t inputWidth,
+    const int64_t outputHeight,
+    const int64_t outputWidth,
+    T* output) {
+  for (int linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < numInputElements;
+       linearIndex += blockDim.x * gridDim.x) {
+    int c = (linearIndex / inputWidth / inputHeight) % numChannels;
+    int n = linearIndex / inputWidth / inputHeight / numChannels;
+    input += (n * numChannels + c) * outputHeight * outputWidth;
+    int maxind = indices[linearIndex];
+    output[linearIndex] = input[maxind];
+  }
+}
+
+template <typename T>
+__global__ void MaxUnpooling3d_backward_kernel(
```

Comment:
I found out that I can't use array pointers as usual, as those pointers reference CPU memory - so I ended up passing in a ton of args, which is clearly kinda janky. Is there an idiomatic way to pass in an array into a CUDA kernel in ATen? There seem to be helper functions in THNN.